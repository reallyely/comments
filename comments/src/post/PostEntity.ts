import { CommentEntity } from "@/modules/Comment/CommentEntity"
import { AuthorEntity } from "@/modules/Author/AuthorEntity"
import { AggregateRoot } from "@/core/Aggregate";

interface PostProps {
  content: string;
  author: AuthorEntity
  comments: Array<CommentEntity>
  dateCreated?: Date
}

interface PostPropsFromView {
  content: string;
  author: AuthorEntity
}

export class PostEntity extends AggregateRoot<PostProps> {
  public content: string;
  public author: AuthorEntity
  public comments: Array<CommentEntity>
  public dateCreated: Date
  private constructor(obj: PostProps) {
    super(obj)
    this.content = obj.content
    this.author = obj.author
    this.comments = obj.comments
    this.dateCreated = new Date()
  }
  public static isValid(obj: any) {
    const mandatoryFields = ["content", "author"]
    return mandatoryFields.filter((key) => Object.keys(obj).includes(key)).length >= mandatoryFields.length
  }
  public static createFromJSON(json: string): PostEntity {
    const parsed = JSON.parse(json)
    if (PostEntity.isValid(parsed)) {
      const { content, author, comments } = parsed
      return new PostEntity({ content, author, comments });
    }
    return PostEntity.createEmpty()
  }
  public static fromForm(obj: PostPropsFromView): PostEntity {
    if (PostEntity.isValid(obj)) {
      return new PostEntity({ ...obj, comments: [CommentEntity.createEmpty()] })
    }
    return PostEntity.createEmpty()
  }
  public static createEmpty(): PostEntity {
    return new PostEntity({ content: "", author: AuthorEntity.createEmpty(), comments: [CommentEntity.createEmpty()] })
  }
  public static create(obj: PostProps): PostEntity {
    return new PostEntity(obj)
  }
}


