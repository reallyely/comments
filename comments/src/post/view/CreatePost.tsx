import { TextField, Stack, Container, Button, Box, Card } from "@mui/material";
import { SxProps } from "@mui/system";
import { ChangeEventHandler, PropsWithChildren, useState } from "react";
import { AuthorEntity } from "../mapping/AuthorEntity";

const containerStyles: SxProps = {
  /* Auto layout */
  padding: "16px",
  background: "#FFFFFF",
  /* eFuse Light/UI Colors/GlobalStroke */
  border: "1px solid #CED7E7",
  borderRadius: "10px",
  boxShadow: "none"
}
const inputStyles: SxProps = {
  fontWeight: 500,
  fontSize: "14px",
  borderColor: "red",
  color: "#0A0C1299"
}
const submitStyles: SxProps = {
  background: "#006CFA",
  borderRadius: "8px",
  textTransform: "unset",
  fontSize: "12px",
  fontWeight: 600
}
interface CreatePostProps {
  postContent?: string
  currentUser?: AuthorEntity
  handleUpdatePostContent?(postContent: string): void
  handleCreatePost?(postContent: string): void
}
export function CreatePost(props: PropsWithChildren<CreatePostProps>) {
  const [postContent, updatePostContent] = useState(props.postContent || "")

  const handleUpdatePostContent: ChangeEventHandler<HTMLInputElement> = (event) => {
    if (props.handleUpdatePostContent) {
      props.handleUpdatePostContent(event.target.value)
    }
    updatePostContent(event.target.value)
  }
  const submit = () => {
    if (props.handleCreatePost) {
      props.handleCreatePost(postContent)
    }
    updatePostContent("")
  }
  return <Card sx={
    containerStyles
  }>
    <Stack sx={{ gap: "16px" }}>
      <TextField
        sx={inputStyles}
        value={postContent}
        label="What's on your mind?"
        type="text"
        multiline
        onChange={handleUpdatePostContent}
        onKeyUp={(e) => {
          if (e.key === "Enter") {
            submit()
          }
        }}
      />
      <Stack direction="row" justifyContent="flex-end">
        <Button sx={submitStyles} variant="contained" type="submit" onClick={submit} >Post</Button>
      </Stack>
    </Stack >
  </Card>
}