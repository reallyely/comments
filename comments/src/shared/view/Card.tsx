import { CardProps, Card as MuiCard, SxProps } from "@mui/material";
import { PropsWithChildren } from "react";
const containerStyles: SxProps = {
  padding: "16px",
  gap: "16px",
  background: "#FFFFFF",

  border: "1px solid #CED7E7",
  borderRadius: "10px",
  boxShadow: "none"
}

export function Card(props: PropsWithChildren<CardProps>) {
  return <MuiCard sx={
    containerStyles
  }>
    {props.children}
  </MuiCard>

}
