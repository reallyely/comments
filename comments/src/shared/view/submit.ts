import { KeyboardEventHandler } from "react"

export const submitOnEnter = (submitFn: any): KeyboardEventHandler => (e) => {
  if (e.key === "Enter") {
    submitFn(e)
  }
}