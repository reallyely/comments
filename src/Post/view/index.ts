export * from "./CreatePost"
export * from "./Post"
export * from "./Posts"