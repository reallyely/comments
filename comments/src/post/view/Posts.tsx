import { Post } from "@/post/view"
import { PostEntity } from "../mapping/PostEntity"
interface PostsProps {
  posts: Array<PostEntity>
}
export function Posts(props: PostsProps) {
  return <div>Posts
    {props.posts.map((props) => <Post {...props} key={props.id} />)}
  </div>
}