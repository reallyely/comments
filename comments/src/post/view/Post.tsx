import { AuthorEntity } from "@/modules/Author/AuthorEntity"
import { Author } from "@/modules/Author/Author"
import { PostEntity } from "../PostEntity"
import { Card } from "@/components/Card"
import { Stack, Typography } from "@mui/material"

interface PostProps {
  author: AuthorEntity
  content: string
  dateCreated: Date
}
export function Post(props: PostProps) {
  return <Card>
    <Stack sx={{ gap: "16px" }}>
      <Author
        portraitSource={props.author.portrait.src}
        name={props.author.name}
        postCreated={props.dateCreated}
      ></Author>
      <Typography variant="body1" sx={{ fontFamily: "Open Sans", fontSize: "12px" }}>{props.content}</Typography>
    </Stack>
  </Card>
}