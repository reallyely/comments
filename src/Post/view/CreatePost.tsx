import { Box, Button, Container, Stack, TextField } from "@mui/material";
import { ChangeEventHandler, PropsWithChildren, useState } from "react";

import { Card } from "@/modules/shared/view/Card"
import { SxProps } from "@mui/system";
import { submitOnEnter } from "@/components/submit";

const inputStyles: SxProps = {
  fontWeight: 500,
  fontSize: "14px",
  borderColor: "red",
  color: "#0A0C1299",
  borderRadius: "8px"
}
const submitStyles: SxProps = {
  background: "#006CFA",
  borderRadius: "8px",
  textTransform: "unset",
  fontSize: "12px",
  fontWeight: 600
}
interface CreatePostProps {
  postContent?: string
  handleUpdatePostContent?(postContent: string): void
  handleCreatePost?(postContent: string): void
}
export function CreatePost(props: PropsWithChildren<CreatePostProps>) {
  const [postContent, updatePostContent] = useState(props.postContent || "")

  const handleUpdatePostContent: ChangeEventHandler<HTMLInputElement> = (event) => {
    if (props.handleUpdatePostContent) {
      props.handleUpdatePostContent(event.target.value)
    }
    updatePostContent(event.target.value)
  }
  const submit = () => {

    if (props.handleCreatePost && postContent.length > 0) {
      props.handleCreatePost(postContent)
      updatePostContent("")
    }
  }
  return <Card>
    <Stack>
      <TextField
        sx={inputStyles}
        value={postContent}
        label="What's on your mind?"
        type="text"
        multiline
        onChange={handleUpdatePostContent}
        onKeyDown={submitOnEnter(submit)}
      />
      <Stack direction="row" justifyContent="flex-end">
        <Button sx={submitStyles} variant="contained" type="submit" onClick={submit}>Post</Button>
      </Stack>
    </Stack >
  </Card>
}
