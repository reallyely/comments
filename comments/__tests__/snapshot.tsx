import { PostEntity } from '@/modules/Post/PostEntity'
import { render } from '@testing-library/react'
import { CreatePost, Posts } from '@/modules/Post/view/'
// Shim for uid creation in entities
import crypto from "crypto";
Object.defineProperty(global, 'crypto', {
  value: {
    getRandomValues: (arr: any) => crypto.randomBytes(arr.length)
  }
});

describe("UI Matches Figma designs", () => {
  it('Create Post', () => {
    const { container } = render(<CreatePost />)
    expect(container).toMatchSnapshot()
  })
  it('Posts', () => {
    const initialPost = PostEntity.createEmpty()
    const { container } = render(<Posts posts={[initialPost]} />)
    expect(container).toMatchSnapshot()
  })

})
