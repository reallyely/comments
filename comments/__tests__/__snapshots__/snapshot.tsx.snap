// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders CreatePost unchanged 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1slgqai-MuiPaper-root-MuiCard-root"
  >
    <div
      class="css-1ucnu3w-MuiStack-root"
    >
      <div
        class="MuiFormControl-root MuiTextField-root css-pgbb05-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":r0:"
          id=":r0:-label"
        >
          What's on your mind?
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-multiline css-dpjnhs-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <textarea
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r0:"
            style="height: 0px; overflow: hidden;"
          />
          <textarea
            aria-hidden="true"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
            readonly=""
            style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
            tabindex="-1"
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                What's on your mind?
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="css-181c0ax-MuiStack-root"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-10p7b8c-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="submit"
        >
          Post
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
